README.md
# group-signature-client

![](https://github.com/FISCO-BCOS/FISCO-BCOS/raw/master/docs/images/FISCO_BCOS_Logo.svg?sanitize=true)

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![GitHub issues](https://img.shields.io/github/issues/FISCO-BCOS/group-signature-client.svg)](https://github.com/FISCO-BCOS/group-signature-client/issues)
[![All releases](https://img.shields.io/github/release/FISCO-BCOS/group-signature-client.svg)](https://github.com/FISCO-BCOS/group-signature-client/releases/latest)
![](https://img.shields.io/github/license/FISCO-BCOS/group-signature-client) 

群/环签名客户端，通过向[服务端](https://github.com/FISCO-BCOS/group-signature-server)发起请求生成签名，并将签名上链，然后调用群/环签名[预编译合约](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/manual/privacy.html#id17)实现签名的链上认证。

群/环签名客户端和服务端是专门提供给社区用户的FISCO BCOS[隐私模块](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/manual/privacy.html)关于群/环签名的开发示例，架构如下：

![](image/demo.jpg)

## 代码结构

| <div align = left>目录</div>  | <div align = left>说明</div>                                 |
| --------------------------------- | ------------------------------------------------------------ |
| main/java/org/fisco/bcos/groupsig/app | 客户端主逻辑，包括RPC调用和合约调用 |
| main/java/org/fisco/bcos/groupsig/contract | 由智能合约转换的java代码 |
| main/resources/conf               | 客户端配置文件目录|
| main/resources/sol                | 链上部署的智能合约目录|
| main/resources/node               | 节点配置文件目录 |

## 部署依赖

**部署服务端**

群/环签名客户端要访问签名服务，因此需要先部署服务端，详细步骤可参考[操作文档](https://github.com/FISCO-BCOS/group-signature-server)。

**部署FISCO BCOS**

客户端可将群/环签名上链，若要使用链上验证功能，需要部署FISCO BCOS，并开启隐私模块开关。

FISCO BCOS详细搭建步骤，以及隐私模块启用方法可参考[操作文档](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/manual/privacy.html)。

## 编译与配置

依赖部署完毕后，就可以拉取代码，编译并配置客户端。详细步骤如下：

**编译客户端**

```bash
git clone https://github.com/FISCO-BCOS/group-signature-client
# 进入客户端目录
cd group-signature-client
# 编译客户端需安装jdk v1.8及以上，gradle v4.6及以上
./gradlew build
# 编译成功后会生成目标目录dist
```

**配置客户端**

(1) 配置RPC服务：`dist/conf/conn.json`

- ip：服务端IP，默认是127.0.0.1
- port：服务端RPC监听端口，默认是8005
- thread_num：线程数目，压测场景使用

(2) 配置日志：`dist/conf/log4j.properties`,`dist/conf/clog.ini`

- 使用默认配置即可

(3) 配置区块链节点信息：`dist/conf/node`

- 将`dist/conf/`目录下的`config-sample.toml`拷贝成`config.toml`
- 将节点的`nodes/127.0.0.1/sdk`目录下的`ca.crt`、`sdk.crt`和`sdk.key`文件拷贝`dist/conf/node`目录下
- `config-sample.toml`详细配置说明可参考配置中注释

## 使用客户端

配置完客户端后就可以开始体验它的相关功能了。

### 群签名示例

接口说明详见[群签名接口文档](doc/interface.md#群签名接口)，以下命令在dist目录下执行。

**(1)创建群**

创建群组group1，创建群时使用了A类型线性对。

```bash
# 参数说明
# 启动客户端程序：java -cp 'apps/*:lib/*:conf' org.fisco.bcos.groupsig.app.Main
# args[0] : 调用的接口名称
# args[1:]: 调用的接口参数 
java -cp 'apps/*:lib/*:conf' org.fisco.bcos.groupsig.app.Main create_group 'group1' '123' '{"linear_type":"a", "q_bits_len":256, "r_bits_len":256}'
```

**(2)群成员加入**

在group1中加入群成员member1。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main join_group 'group1' 'member1'
```

**(3) 生成群签名**

生成群group1中成员member1对信息“hello”的群签名。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main group_sig 'group1' 'member1' 'hello'
```

**(4) 产生群签名，将签名信息上链**

首先生成群group1成员member1对信息“hello”的群签名，然后使用[AMOP](https://github.com/FISCO-BCOS/Wiki/tree/master/AMOP%E4%BD%BF%E7%94%A8%E6%8C%87%E5%8D%97)将群签名信息写到链上，返回合约地址。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main deploy_group_sig 'group1' 'member1' 'hello'
```

**(5) 链上验证群签名**

示例：验证部署于地址为0xd6c8a04b8826b0a37c6d4aa0eaa8644d8e35b79f合约处的群签名有效性（返回群签名验证结果true/false）。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main group_sig_verify '0xd6c8a04b8826b0a37c6d4aa0eaa8644d8e35b79f'
```

### 环签名示例

接口说明详见[环签名接口文档](doc/interface.md#环签名接口)。

**(1) 初始化环**

调用127.0.0.1 8005端口的群签名服务(群签名RPC服务部署方法参考[群/环签名RPC](https://github.com/FISCO-BCOS/group-signature-server/tree/dev-2.0) )，初始化环ring1(环成员公/私钥长度默认为1024bits)。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main setup_ring 'ring1'
```

**(2) 加入环成员**

在环ring1中加入一个环成员，可重复执行该命令来生成多个环成员。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main join_ring 'ring1'
```

**(3) 生成环签名**

为环ring1的位置0处的环成员生成对消息“hello”环签名，假设当前的环成员大于4。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main ring_sig 'hello' 'ring1' '0' '4'
```

**(4) 生成环签名，并将签名信息上链**

首先生成环ring1位置0成员对消息“hello”的环签名，然后使用[AMOP](https://github.com/FISCO-BCOS/Wiki/tree/master/AMOP%E4%BD%BF%E7%94%A8%E6%8C%87%E5%8D%97)将环签名信息写到链上，返回合约地址。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main deploy_ring_sig 'hello' 'ring1' '0' '4'
```

**(5) 环签名信息链上验证**

示例：验证部署于地址为0x2426edaa1173f65cd7d62c93c935bfde329d247c处合约中的环签名有效性，返回验证结果。

```bash
java -cp 'conf/:lib/*:apps/*' org.fisco.bcos.groupsig.app.Main ring_sig_verify '0x2426edaa1173f65cd7d62c93c935bfde329d247c'
```

### 注意事项

- 环签名限制

由于环签名的签名长度以及签名/验证时间与环成员数目呈线性关系，为防止超gas，每个环的环成员数量不能超过32个。

- 群签名算法签名和验证时间与群成员数目无关，具有良好的可扩展性。

- 群签名算法签名长度与线性对参数设置有关，不同线性对的公私钥长度和签名长度（字节）如下：

| 线性对参数类型                                                   | 私钥 |   公钥 | 签名 |
| ------------------------------------------------------------ | ------------ |------------ |------------ |
|**A：**{"linear_type":"a","q_bit_len":256,"r_bit_len":256} |912 |2700 |1602 |
|**A1：**{"linear_type":"a_one","order":512} |1293 |5180 |3102 |
|**E：**{"linear_type":"e","q_bits_len":1024,"r_bits_len":160}| 905| 4148| 2328|
|**F：**{"linear_type":"f","bit_len":256}  |653 |2358 | 1584|

## 接口说明

详见[接口文档](doc/interface.md)。

## 贡献代码

- 我们欢迎并非常感谢您的贡献，请参阅[代码贡献流程](CONTRIBUTING.md)。
- 如项目对您有帮助，欢迎star支持！

## 加入社区

**FISCO BCOS开源社区**是国内活跃的开源社区，社区长期为机构和个人开发者提供各类支持与帮助。已有来自各行业的数千名技术爱好者在研究和使用FISCO BCOS。如您对FISCO BCOS开源技术及应用感兴趣，欢迎加入社区获得更多支持与帮助。

![](https://raw.githubusercontent.com/FISCO-BCOS/LargeFiles/master/images/QR_image.png)

## License

![license](https://img.shields.io/github/license/FISCO-BCOS/group-signature-client.svg)

group-signature-client的开源协议为[APACHE LICENSE 2.0](http://www.apache.org/licenses/). 详情参考[LICENSE](./LICENSE)。


Changelog.md
### v2.0.0

(2020-3-18)

**新特性**

- 支持调用服务端RPC接口，实现创建群、加入群、以及生成群签名等功能
- 支持调用服务端RPC接口，实现初始化环、加入环、以及生成环签名等功能
- 支持与FISCO BCOS节点通讯，将群/环签名以及签名信息上链
- 支持调用FISCO BCOS预编译合约接口，实现群/环签名的链上验证

### v3.0.0

(2022-5-18)

**新特性**

- 支持调用服务端RPC接口，实现创建群、加入群、以及生成群签名等功能
- 支持调用服务端RPC接口，实现初始化环、加入环、以及生成环签名等功能
- 支持java-sdk与FISCO BCOS节点通讯，将群/环签名以及签名信息上链
- 支持java-sdk调用FISCO BCOS预编译合约接口，实现群/环签名的链上验证


sig_client/src/main/executive/shell/runSigService.sh
#!/bin/bash
dirpath="$(cd "$(dirname "$0")" && pwd)"

SigService="$dirpath/sig_client"
chmod +x ${SigService}

# @function: output log with red color (error log)
# @param: content: error message
function LOG_ERROR()
{
    local content=${1}
    echo -e "\033[31m"${content}"\033[0m"
}

# @function: output information log
# @param: content: information message
function LOG_INFO()
{
    local content=${1}
    echo -e "\033[34m"${content}"\033[0m"
}

# @function: execute command
# @params: 1. command: command to be executed
function execute_cmd()
{
    local command="${1}"
    eval "${command}"
    local ret=$?
    if [ $ret -ne 0 ];then
        LOG_ERROR "execute command ${command} FAILED"
        exit 1
    else
        LOG_INFO "execute command ${command} SUCCESS"
    fi
}

# @function: deploy contract, and write group signature to blockchain
# @params: 1. config_file: config file path, config (ip, port)
#          2. group_name: group_name used to generate group sig 
#          3. member_name: member that want to genrate group sig, 
#             and deploy the sig to blockchain
#          4. message: plain text of the signature

function deploy_group_sig()
{
    local config_file="${1}"
    local group_name="${2}"
    local member_name="${3}"
    local message="${4}"

    local command=${SigService}" ${config_file} deploy_group_sig user.jks 123456 123456 ${group_name} ${member_name} ${message}"
    execute_cmd "${command}"
}


###deploy ring sig contract
# @function: deploy contract, and write ring signature to blockchain
# @params: 1. config_file: file path of configuration(used to config (ip, port))
#          2. message: plain text of the signature
#          3. ring_name: specified ring the signature belongs to
#          4. mem_pos: position of specified member in specified ring to 
#             deploy its signature to blockchain
#          5. ring_size: option,
#             number of public keys of all used to generate ring signature 
function deploy_ring_sig()
{
    local config_file="${1}"
    local message="${2}"
    local ring_name="${3}"
    local mem_pos="${4}"
    local ring_size="4"
    if [ $# -ge 5 ];then
        ring_size="${5}"
    fi
    local command=${SigService}" ${config_file} deploy_ring_sig user.jks 123456 123456 ${message} ${ring_name} ${mem_pos} ${ring_size}"
    execute_cmd "${command}"
}

# @function: verify specified group signature on blockchain is valid or not
# @params: 1. config_file: file path of (ip, port) configuration
#          2. contract_address: contract address related to some group signature
#          3. stress_test: (1) "0": callback on-chain verification only once
#                          (2) "1": callback on-chian verification continually
#                              (for stress testing) 
function eth_group_verify()
{
    local config_file="${1}"
    local contract_address="${2}"
    local stress_test="0"
    if [ $# -ge 3 ];then
        stress_test="${3}"
    fi
    local command="${SigService} ${config_file} group_sig_verify user.jks 123456 123456 ${contract_address} ${stress_test}"
    execute_cmd "${command}"
}

# @function: verify specified ring signature on blockchain is valid or not
# @params: 1. config_file: file path of (ip, port) configuration
#          2. contract_address: contract address related to some ring signature
#          3. stress_test: (1) "0": callback on-chain verification only once
#                          (2) "1": callback on-chian verification continually
#                              (for stress testing) 
function eth_ring_verify()
{
    local config_file="${1}"
    local contract_address="${2}"
	local stress_test="0"
	if [ $# -ge 3 ];then
		stress_test="${3}"
	fi
    local command="${SigService} ${config_file} ring_sig_verify user.jks 123456 123456 ${contract_address} ${stress_test}"
    echo "command = "${command}
    execute_cmd "${command}"
}

######*******group sig related RPC**********####
# @function: create group with specified group_name
#           (generate public keys, params and group manager private key)
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name
#          3. gm_pass: password of group manager(is reserved now, can be every string)
#          4. pbc_param: specified pbc linear pair used to generate group param, 
#             default is A type linear pair, examples:
#             (1) A type linear pair: 
#               '"{\"linear_type\": \"a\", \"q_bits_len\": , \"r_bits_len\": 256, \"q_bits_len\":256}"'
#             (2) A1 type linear pair: 
#               '"{\"linear_type\":\"a_one\", \"order\":512}"'
#             (3) E type linear pair:
#              '"{\"linear_type\":\"e\", \"q_bits_len\": 512, \"r_bits_len\":512}"'
#             (4) F type linear pair:
#              '"{\"linear_type\":\"f\", \"bit_len\": 256}"'
function create_group()
{
    local config_file="${1}"
    local group_name="${2}"
    local gm_pass="${3}"
    local pbc_param=""
    if [ $# -ge 4 ];then
        pbc_param="${4}"
    fi
    local command="${SigService} ${config_file} create_group ${group_name} ${gm_pass} ${pbc_param}"
    LOG_INFO "command=${command}"
    execute_cmd "${command}"
}

##function: join group
# @function: group manager generate private key && cert for group members
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name of the member wants to join in
#          3. member_name: member id of the one wants to obtain group private key and cert
function join_group()
{
    local config_file="${1}"
    local group_name="${2}"
    local member_name="${3}"
    local command="${SigService} ${config_file} join_group ${group_name} ${member_name}"
    execute_cmd "${command}"
}

# @function: generate signature with bbs04 group sig algorithm 
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name of the signature belongs to 
#          3. member_name: member name of the signature belongs to
#          4. message: plain text 
function group_sig()
{
    local config_file="${1}"
    local group_name="${2}"
    local member_name="${3}"
    local message="${4}"
    local stress_test="0"
    if [ $# -ge 5 ];then
        stress_test="${5}"
    fi
    local command="${SigService} ${config_file} group_sig ${group_name} ${member_name} ${message} ${stress_test}"
    execute_cmd "${command}"
}

# @function: verify specified signature
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group name: group name the signature belongs to
#          3. sig: signature
#          4. message: plain text
#          5. stress test: optional
#             (1)"0": verify only once
#             (2) "1": verify continually, used for stress testing
function group_verify()
{
    local config_file="${1}"
    local group_name="${2}"
    local sig="${3}"
    local message="${4}"
    local stress_test="0"
    if [ $# -ge 5 ];then
        stress_test="${5}"
    fi
    local command="${SigService} ${config_file} group_verify ${group_name} ${sig} ${message} ${stress_test}"
    LOG_INFO "GROUP VERIFY COMMAND = ${command}"
    execute_cmd "${command}"
}

##open_cert
# @function: get cert according to given signature
#            (only group manager can calculate the cert)
#            (generally used in regulation cases)
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name of specified signature belongs to
#          3. sig: signature to trace its signer
#          4. message: plain text
#          5. gm_pass: password of group manager(reserved now, can be every string)
function open_cert()
{
    local config_file="${1}"
    local group_name="${2}"
    local sig="${3}"
    local message="${4}"
    local gm_pass="${5}"
    local command="${SigService} ${config_file} open_cert ${group_name} ${sig} ${message} ${gm_pass}"
    execute_cmd "${command}"
}

# @function: get public key of specified group
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name 
function get_public_info()
{
    local config_file="${1}"
    local group_name="${2}"
    local command="${SigService} ${config_file} get_public_info ${group_name}"
    execute_cmd "${command}"
}

# @function: get private key of group manager
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name of the group manager belongs to
#          3. gm_pass: password of group manager(is reserved now, can be every string)
function get_gm_info()
{
    local config_file="${1}"
    local group_name="${2}"
    local gm_pass="${3}"
    local command="${SigService} ${config_file} get_gm_info ${group_name} ${gm_pass}"
    execute_cmd "${command}"
}

# @function: get private key of specified group member
# @params: 1. config_file: file path of (ip, port) configuration
#          2. group_name: group name the member belongs to
#          3. member_name: member name
#          4. pass: password of the member(is reserved now, can be every string)
function get_member_info()
{
    local config_file="${1}"
    local group_name="${2}"
    local member_name="${3}"
    local pass="${4}"
    local command="${SigService} ${config_file} get_member_info ${group_name} ${member_name} ${pass}"
    execute_cmd "${command}"
}

############Ring Sig related RPC#######

# @function: generate params for specified ring 
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
function setup_ring()
{
    local config_file="${1}"
    local ring_name="${2}"
    local bit_len="1024"
    if [ $# -ge 3 ];then
        bit_len="${3}"
    fi
    local command="${SigService} ${config_file} setup_ring ${ring_name} ${bit_len}"
    execute_cmd "${command}"
}

# @function: generate private key for member of specified ring
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
function join_ring()
{
    local config_file="${1}"
    local ring_name="${2}"
    local command="${SigService} ${config_file} join_ring ${ring_name}"
    execute_cmd "${command}"
}

# @function: generate signature with linkable ring sig
#            algorithm for specified ring member
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
#          3. member_pos: postion of the member to generate signature
#          4. message: plain text
#          5. ring_size: ring size of the member used to generate signature
#          6. stress_test: optional
#          (1) "0": only generate signature once
#          (2) "1": generate signature continually(for stress testing)
function ring_sig()
{
    local config_file="${1}"
    local ring_name="${2}"
    local member_pos="${3}"
    local message="${4}"
    local ring_size="4"
    local stress_test="0"
    if [ $# -ge 5 ];then
        ring_size="${5}"
    fi
    if [ $# -ge 6 ];then
        stress_test="${6}"
    fi
    local command="${SigService} ${config_file} ring_sig ${message} ${ring_name} ${member_pos} ${ring_size} ${stress_test}"
    execute_cmd "${command}"
}

# @function: verify given signature is valid or not
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
#          3. signature: linkable ring signature
#          4. message: plain text of given signature
#          5. stress_test: optional
#          (1) "0": only verify signature once
#          (2) "1": verify signature continually(for stress testing)
function ring_verify()
{
    local config_file="${1}"
    local ring_name="${2}"
    local sig="${3}"
    local message="${4}"
    local stress_test="0"
    if [ $# -ge 5 ];then
        stress_test="${5}"
    fi
    local command="${SigService} ${config_file} ring_verify ${ring_name} ${sig} ${message} ${stress_test}"
    execute_cmd "${command}"
}

# @function: get ring params according to given ring_name
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring_name
function get_ring_param()
{
    local config_file="${1}"
    local ring_name="${2}"
    local command="${SigService} ${config_file} get_ring_param ${ring_name}"
    execute_cmd "${command}"
}

# @function: get public key of specified member
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
#          3. member_pos: position of the public key
function get_ring_public_key()
{
    local config_file="${1}"
    local ring_name="${2}"
    local member_pos="${3}"
    local command="${SigService} ${config_file} get_ring_public_key ${ring_name} ${member_pos}"
    execute_cmd "${command}"
}

# @function: get private key of specified ring member
# @params: 1. config_file: file path of (ip, port) configuration
#          2. ring_name: ring name
#          3. member_pos: position of the member
function get_ring_private_key()
{
    local config_file="${1}"
    local ring_name="${2}"
    local member_pos="${3}"
    local command="${SigService} ${config_file} get_ring_private_key ${ring_name} ${member_pos}"
    execute_cmd "${command}"
}

# @function: get time interval
# @params: 1. start: start time
#          2. end: end time
function get_time()
{
    local start="${1}"
    local end="${2}"
    local start_s=$(echo $start | cut -d '.' -f 1)
    local start_ns=$(echo $start | cut -d '.' -f 2) 
    local end_s=$(echo $end | cut -d '.' -f 1)
    local end_ns=$(echo $end | cut -d '.' -f 2)

    echo "start:"${start}
    echo "end:"${end}

    local time=$(( ( 10#$end_s - 10#$start_s ) * 1000 + ( 10#$end_ns / 1000000 - 10#$start_ns / 1000000 ) )) 
    echo "${time} ms"
}

# @function: update ring signature of specified contract address 
# @params: 1. config_file: file path of (ip, port) configuration
#          2. contract_addr: contract address whose signature should be updated
#          3. message: plain text of the new signature
#          4. ring_name: ring name
#          5. mem_pos: position of the member to generate new signature
#          6. ring_size: optional
#             ring size of the member used to generate signature(default is 4) 
function update_ring_sig_data()
{
    local config_file="${1}"
    local contract_addr="${2}"
    local message="${3}"
    local ring_name="${4}"
    local mem_pos="${5}"
    local ring_size="4"
    if [ $# -ge 6 ];then
        ring_size="${6}"
    fi
    local command="${SigService} ${config_file} update_ring_sig_data user.jks 123456 123456 ${contract_addr} ${message} ${ring_name} ${mem_pos} ${ring_size}"
    execute_cmd "${command}"
}

# @function: update group signature of specified contract address
# @params: 1. config_file: file path of (ip, port) configuration
#          2. contract_addr: contract address whose signature should be updated
#          3. message: plain text of the new signature
#          4. group_name: group of the new signature
#          5. mem_name: the signer of the signature
function update_group_sig_data()
{
    local config_file="${1}"
    local contract_addr="${2}"
    local group_name="${3}"
    local mem_name="${4}"
    local message="${5}"
    local command="${SigService} ${config_file} update_group_sig_data user.jks 123456 123456 ${contract_addr} ${group_name} ${mem_name} ${message}"
    execute_cmd "${command}"
}
####group sig example
#create_group "conn.json" "group4" "123" '"{\"linear_type\":\"f\", \"bit_len\":256}"'
#join_group "conn.json" "group4" "member1"
#group_sig "conn.json" "group4" "member1" "hello"
#deploy_group_sig "conn.json" "group4" "member1" "hello"
#true means that verify succ
#eth_group_verify "conn.json" "0x1af99bf7d62bf6629459d47399358b23c19fcbd2"
#update_group_sig_data "conn.json" "0x1af99bf7d62bf6629459d47399358b23c19fcbd2" "group4" "test1" "hello" 


####ring sig example
#setup_ring "conn.json" "ring1"
#join_ring "conn.json" "ring1"
#join_ring "conn.json" "ring1"
#join_ring "conn.json" "ring1"
#join_ring "conn.json" "ring1"
#ring_sig "conn.json" "ring1" "0" "hello" "4"
#1 represents that verify succ
#deploy_ring_sig "conn.json" "hello" "ring1" "0" "4"
#eth_ring_verify "conn.json" "0x418d9da6382993e6955135db364c7dba88e483a4"
#update_ring_sig_data "conn.json" "0x418d9da6382993e6955135db364c7dba88e483a4" "hello" "ring1" "1" "4"
